### PR TITLE
Add Resource TopicInlinePolicy

### PR DESCRIPTION
**Pull request by shzShenghao**
*Description of changes:*
Handler code for AWS::SNS::TopicInlinePolicy. The code was created using cfn init, and implemented the appropriate behavior using AWS SDKs.

Here is the code coverage:
Screen Shot 2023-05-22 at 11 17 44 AM

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/UpdateHandler.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            final Logger logger) {
+        final ResourceModel model = request.getDesiredResourceState();
+        final String policy = getPolicyDocument(request);
+        final CallbackContext callbackContext = progress.getCallbackContext();
+        final String topicArn = model.getTopicArn();
+        return proxy.initiate("AWS-SNS-TopicInlinePolicy::Update", proxyClient, model, callbackContext)
+                .translateToServiceRequest((resourceModel) -> Translator.translateToSetRequest(topicArn, policy))
+                .makeServiceCall((awsRequest, client) -> {
+                    SetTopicAttributesResponse response = proxyClient.injectCredentialsAndInvokeV2(awsRequest, client.client()::setTopicAttributes);
+                    logger.log ("Update in progress");
```
Comment (hlazrv) [inline aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/UpdateHandler.java:74]: Here as well to include request.getStackId())

File: aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/CreateHandler.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            final Logger logger) {
+        final ResourceModel model = request.getDesiredResourceState();
+        final CallbackContext callbackContext = progress.getCallbackContext();
+        final String policy = getPolicyDocument(request);
+        final String topic = model.getTopicArn();
+        return proxy.initiate("AWS-SNS-TopicInlinePolicy::Create", proxyClient, model, callbackContext)
+                .translateToServiceRequest((resourceModel) -> Translator.translateToSetRequest(topic, policy))
+                .makeServiceCall((awsRequest, client) -> {
+                    SetTopicAttributesResponse response = proxyClient.injectCredentialsAndInvokeV2(awsRequest, client.client()::setTopicAttributes);
+                    logger.log ("Create in progress");
```
[the remark this answers is not in the document]
Comment (shzShenghao) [reply aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/CreateHandler.java:76]: Line 38. There is the log for StackID.

File: aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/ReadHandler.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+
+        final ResourceModel model = request.getDesiredResourceState();
+
+        if (model == null || StringUtils.isNullOrEmpty(model.getTopicArn())) {
+            return ProgressEvent.failed(model, callbackContext, HandlerErrorCode.InvalidRequest, EMPTY_POLICY_AND_TOPICARN_ERROR_MESSAGE );
+        }
+
+        logger.log(String.format("StackId: %s, ClientRequestToken: %s] Calling Read TopicInlinePolicy", request.getStackId(), request.getClientRequestToken()));
+        if (!doesTopicPolicyExist(proxyClient, request, model)) {
```
Comment (ceskander-aws) [inline aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/ReadHandler.java:35]: does the native code call the GetAPI in your CreateHandler? if is does not you will need to implement a soft failure for adding a read permission inside your create handler. If you need an example please dm me

File: aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/DeleteHandler.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+        return ProgressEvent.progress(resourceModel, callbackContext)
+                .then(progress -> {
+                    if (!doesTopicPolicyExist(proxyClient, request, resourceModel) && !callbackContext.isPropagationDelay()) {
+                        return ProgressEvent.failed(null, callbackContext, HandlerErrorCode.NotFound, noSuchPolicyErrorMessage(resourceModel));
+                    }
+                    return progress;
+                })
+                .then(progress -> Delete(proxy, proxyClient, request, progress, logger))
+                .then(progress -> {
```
Comment (ceskander-aws) [inline aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/DeleteHandler.java:41]: is this delete PropagationNeeded?

File: aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/ListHandler.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+public class ListHandler extends BaseHandlerStd {
+
+    public ProgressEvent<ResourceModel, CallbackContext> handleRequest(
+            final AmazonWebServicesClientProxy proxy,
+            final ResourceHandlerRequest<ResourceModel> request,
+            final CallbackContext callbackContext,
+            final ProxyClient<SnsClient> proxyClient,
+            final Logger logger) {
+        ResourceModel resourceModel = request.getDesiredResourceState();
+        return ProgressEvent.failed(resourceModel, callbackContext, HandlerErrorCode.InvalidRequest,
```
Comment (ceskander-aws) [inline aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/ListHandler.java:20]: see your DM for example of how to implement LIST handler when resource does not support LIST operatio

**Review by mangeche — APPROVED** [review]

File: aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/ReadHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (shzShenghao) [reply aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/ReadHandler.java]: Added

File: aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/CreateHandler.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        return ProgressEvent.progress(resourceModel, callbackContext)
+                .then(progress -> {
+                    if (doesTopicPolicyExist(proxyClient, request, resourceModel) && !callbackContext.isPropagationDelay()) {
+                        return ProgressEvent.failed(null, callbackContext, HandlerErrorCode.AlreadyExists, alreadyExistsErrorMessage(resourceModel));
+                    }
+                    return progress;
+                })
+                .then(progress -> Create(proxy, proxyClient, request, progress, logger))
+                .then(progress -> {
+                    if (!callbackContext.isPropagationDelay()) {
```
[the remark this answers is not in the document]
Comment (shzShenghao) [reply aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/CreateHandler.java:48]: Yeah, discussed with SNS team. set it to 30s

File: aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/ListHandler.java (hunk 5 above)
Comment (shzShenghao) [reply aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/ListHandler.java:21]: Changed

File: aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/DeleteHandler.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            final ResourceHandlerRequest<ResourceModel> request,
+            final CallbackContext callbackContext,
+            final ProxyClient<SnsClient> proxyClient,
+            final Logger logger) {
+
+        this.logger = logger;
+        ResourceModel resourceModel = request.getDesiredResourceState();
+
+        if (resourceModel == null || StringUtils.isNullOrEmpty(resourceModel.getTopicArn())) {
+            return ProgressEvent.failed(resourceModel, callbackContext, HandlerErrorCode.InvalidRequest, "TopicArn can not be empty" );
```
Comment (shih) [inline aws-sns-topicinlinepolicy/src/main/java/software/amazon/sns/topicinlinepolicy/DeleteHandler.java:28]: How about making `TopicArn can not be empty` a static string, and refer in both code and UT?

File: aws-sns-topicinlinepolicy/src/test/java/software/amazon/sns/topicinlinepolicy/CreateHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,605 @@
...
+                .build();
+
+        final ResourceHandlerRequest<ResourceModel> request = ResourceHandlerRequest.<ResourceModel> builder()
+                .desiredResourceState(model)
+                .build();
+
+        InvalidParameterException exception = InvalidParameterException.builder()
+                .awsErrorDetails(AwsErrorDetails.builder()
+                        .errorCode("InvalidParameter")
+                        .errorMessage("Invalid parameter: Policy Error: PrincipalNotFound").build())
```
Comment (shih) [inline aws-sns-topicinlinepolicy/src/test/java/software/amazon/sns/topicinlinepolicy/CreateHandlerTest.java:374]: How about making `Invalid parameter: Policy Error: PrincipalNotFound` as static, and referring in both code and tests?

**Review by shih — APPROVED** [review]

**Pull request merged**